### DISCUSSION TITLE
Composing a RHF form with Child Components

### DISCUSSION THREAD

**Discussion by chandu** [react-hook-form/react-hook-form#2503] [Q&A]
HI,
I was trying to get my head around on how do I compose a RHF form with child components.
The code I am referring to along with working example is @ https://codesandbox.io/s/muddy-sunset-0trgj

Let's say I have my data models {DeliveryData, DeliveryFileDetails} as below:
```
interface DeliveryData {
  mainFile?: DeliveryFileDetails,
  rejectFile?: DeliveryFileDetails
}


interface DeliveryData {
  mainFile?: DeliveryFileDetails,
  rejectFile?: DeliveryFileDetails
}
```
I want to structure my form as below
```
 DeliveryDataComponent -> Parent Component capturing main and reject file and defines the form.
      DeliveryFileInfo -> Child component that renders form inputs with the model as DeliveryData
```

If you see mainFile and rejectFile are structurally same (capturing DeliveryFileDetails), so I have extracted the inputs to capture this data as it's own component(DeliveryFileInfo) and managed to make it work as below:

```
<DeliveryDataComponent>
        <DeliveryFileInfo propName="mainFile." control={formMethods.control} errors={errors?.mainFile} touched={formState.touched.mainFile} dirtyFields={formState.dirtyFields.mainFile} />
        <DeliveryFileInfo propName="rejectFile." control={formMethods.control} errors={errors?.rejectFile} touched={formState.touched.rejectFile} dirtyFields={formState.dirtyFields.rejectFile} />
</DeliveryDataComponent>
       
```

If you look at the `DeliveryFileInfo` I had to leak the property name (`mainFile.`, `rejectFile.`) as props to the child component and this is brittle. I am letting the child component know that path of the properties it's input is rendering/managing etc. Also I had define the `control` as `Control<any>` otherwise Typescript was complaining about the name prop I was assigning to the controller.

Is there any better way to achieve what I am trying to do, basically I want to have the values/state of the RHF form composed of data/inputs from child components without leaking information of project paths?

Or can this be a feature, something like 
```
 const formMethods:UseFormMethods<DeliveryData> = useForm<DeliveryData>({
 .....
})

const mainFileFormMethods:UseFormMethods<DeliveryFileInfo> = formMethods.slicedForm('mainFile')
.
.
.
.
<DeliveryFileInfo formMethods={mainFileFormMethods} />
```

**Comment by bluebill1049** [comment]
would you like to give an example in simple code? (CSB)

**Comment by chandu** [comment]
@bluebill1049 I have simplified the model I am trying to bind to my form and created new Sandbox @ 
https://codesandbox.io/s/dazzling-currying-hix5h?file=/src/Thampu.tsx

Basically I want to create components that capture a part of the model.

e.g: Let's say we have component that captures a Department's attributes (name). I want to be able to reuse this component across different forms that might be bound to different models.

```
interface Department {
  name: string
}

interface Employee {
  id: string;
  employeeDepartment: Department
}

interface Person {
  id: string;
  personDepartment: Department
}
```
Based on the models above, I would like to have one component that captures Department's name and be able to build two forms PersonForm, EmployeeForm   both of which bind to Person, Employee models respsectively.

**Reply by bluebill1049** [reply]
maybe you child component should always have a name attribute `="{parentName}.name"`, then you can reuse?

**Reply by chandu** [reply]
@bluebill1049  Thanks. This is what I am doing right now (as depicted in the examples.)
Wanted to see if there is any better way.
Also right now I am ending up sending almost all the UseFormMethods methods as props to the child component.
I understand that I could use useFormContext in the child component, but in that case and also with passing the name attribute, I am making my Department component tightly coupled with the structure of the model of the parent component (in this case either Employee or Person.)

```
const PersonComponent: React.FC = () => {
  const formMethods = useForm<Person>({
    resolver: yupResolver(personValidator),
    mode: "onChange",
    reValidateMode: "onChange"
  });
  const { errors, formState, handleSubmit, getValues, trigger } = formMethods;
  return (
    <form
      onSubmit={handleSubmit((e) => {
        console.log(e);
      })}
    >
      <button>Submit</button>&nbsp;
      <button
        type="button"
        onClick={() => {
          trigger()
            .then((a) => console.log("Is valid?", a))
            .catch((err) => console.error(err));
        }}
      >
        Validate
      </button>
      <DepartmentComponent
        propName="department."
        control={formMethods.control}
        errors={errors?.department}
        touched={formState.touched.department}
        dirtyFields={formState.dirtyFields.department}
      />
      <pre>
        {JSON.stringify(
          {
            values: getValues(),
            errors
          },
          null,
          2
        )}
      </pre>
    </form>
  );
};
```

**Comment by chandu** [comment]
@bluebill1049  Basically I am looking for a solution similar to the `Grouping` proposed @ https://github.com/react-hook-form/react-hook-form/issues/73#issue-454899901
I would like to be able to build a component that captures a slice of the form data but can be reused across multiple forms bound with different property names (like how Department component can be used to bind to `personDepartment` and `employeeDepartment` in my example)

**Comment by genshinw** [comment]
I really want to know do you find a better way to handle the compose in RHF form?

**Reply by Moshyfawn** [reply]
#7354
